arbnode/mel/delayed_message_backlog.go

```diff
@@ -0,0 +1,122 @@
+	Index                       uint64
+	MsgHash                     common.Hash
+	MelStateParentChainBlockNum uint64
+}
+
+// DelayedMessageBacklog is a data structure that holds metadata related to delayed messages that have been SEEN by MEL but not yet READ.
+// This enables verification of delayed messages read from a database against the current Merkle root of the head MEL state. The MEL state
+// also contains compact witnesses of a Merkle tree representing all seen delayed messages. To prove that a delayed message is part of
+// this Merkle tree, this data structure can be used to verify Merkle proofs against the MEL state.
+type DelayedMessageBacklog struct {
+	ctx                          context.Context
+	capacity                     int // 0 = no trimming
+	entries                      []*DelayedMessageBacklogEntry
+	initMessage                  *DelayedInboxMessage
+	finalizedAndReadIndexFetcher func(context.Context) (uint64, error) // nil = no trimming
+}
+
+func NewDelayedMessageBacklog(ctx context.Context, capacity int, finalizedAndReadIndexFetcher func(context.Context) (uint64, error)) *DelayedMessageBacklog {
+	return &DelayedMessageBacklog{
+		ctx:                          ctx,
+		capacity:                     capacity,
+		entries:                      make([]*DelayedMessageBacklogEntry, 0),
+		initMessage:                  nil,
+		finalizedAndReadIndexFetcher: finalizedAndReadIndexFetcher,
+	}
+}
+
+// Add takes values of a DelayedMessageBacklogEntry and adds it to the backlog given the entry succeeds validation. It also attempts trimming of backlog if capacity is reached
+func (d *DelayedMessageBacklog) Add(entry *DelayedMessageBacklogEntry) error {
+	if len(d.entries) > 0 {
+		expectedIndex := d.entries[0].Index + uint64(len(d.entries))
+		if entry.Index != expectedIndex {
+			return fmt.Errorf("message index %d is not sequential, expected %d", entry.Index, expectedIndex)
+		}
+	}
+	d.entries = append(d.entries, entry)
+	return d.clear()
+}
+
+// clear removes from backlog (if exceeds capacity) the entries that correspond to the delayed messages that are both READ and belong to finalized parent chain blocks
+func (d *DelayedMessageBacklog) clear() error {
+	if len(d.entries) <= d.capacity {
+		return nil
+	}
+	if d.finalizedAndReadIndexFetcher != nil {
+		finalizedDelayedMessagesRead, err := d.finalizedAndReadIndexFetcher(d.ctx)
+		if err != nil {
+			log.Error("Unable to trim finalized and read delayed messages from DelayedMessageBacklog, will be retried later", "err", err)
+			return nil // we should not interrupt delayed messages accumulation if we cannot trim the backlog, since its not high priority
+		}
+		if finalizedDelayedMessagesRead > d.entries[0].Index {
+			leftTrimPos := min(finalizedDelayedMessagesRead-d.entries[0].Index, uint64(len(d.entries)))
+			d.entries = d.entries[leftTrimPos:]
+		}
+	}
+	return nil
+}
+
+func (d *DelayedMessageBacklog) Get(index uint64) (*DelayedMessageBacklogEntry, error) {
```

Comment:
Nit, but it's more common to have exported methods first followed by unexported ones